lgc/builder/InOutBuilder.cpp

```diff
@@ -251,19 +114,52 @@ Value *InOutBuilder::CreateReadPerVertexInput(Type *resultTy, unsigned location,
     return emitCall(callName, resultTy, args, {Attribute::ReadOnly, Attribute::WillReturn}, &*GetInsertPoint());
   };
 
+  unsigned oddOffset = 0, evenOffset = 0;
+  auto primType = getPipelineState()->getPrimitiveType();
+  auto provokingVertexMode = getPipelineState()->getRasterizerState().provokingVertexMode;
+  unsigned vertexIndexInt = cast<ConstantInt>(vertexIndex)->getZExtValue();
   Value *result = nullptr;
-  if (parity) {
-    // Odd
-    auto oddRes = readInput(odd);
-    // Even
-    auto evenRes = readInput(even);
-
-    auto primtiveId = readBuiltIn(false, BuiltInPrimitiveId, {}, nullptr, nullptr, "");
-    auto evenV = CreateSRem(primtiveId, getInt32(2));
-    Value *con = CreateICmpEQ(evenV, getInt32(0));
-    result = CreateSelect(con, evenRes, oddRes);
-  } else {
+
+  switch (primType) {
+  case PrimitiveType::TriangleList:
+    vertexIndex = getInt32((vertexIndexInt + 2) % 3); // 0->2, 1->0, 2->1
+    [[clang::fallthrough]];
+    // fall through...
+  case PrimitiveType::Point:
+  case PrimitiveType::LineList:
+  case PrimitiveType::LineStrip:
+    // These are the non-parity modes that just read a single input.
     result = readInput(vertexIndex);
+    break;
+
+  default:
+    // The remaining modes are the parity modes that read two inputs and select between them.
+    // Fix the order.
+    switch (primType) {
+    case PrimitiveType::TriangleFan:
+      oddOffset = provokingVertexMode == ProvokingVertexLast ? 1 : 2;
```

Comment:
evenOffset is missing.
`evenOffset = provokingVertexMode == ProvokingVertexLast ? 2 : 0;`